### ISSUE TITLE
Pre-validation queue for transaction pool

### ISSUE THREAD

**Issue by NikVolf** [paritytech/substrate#6248]
Right now when the node starts to produce a block, it competes for I/o (db) resources with the transaction pool if it is receiving a lot of transactions at the same time.

Most of those transactions received in this time frame are going to be revalidated anyway once new (being produced) block imported, against it, so it is useless work which also cripples block production.

So I have an idea that during block production transaction pool should not validate transactions but instead should put them in the queue and start to validate once the block production is over. 

Happy to discuss

**Comment by bkchr** [comment]
> Right now when the node starts to produce a block, it competes for I/o (db) resources with the transaction pool if it is receiving a lot of transactions at the same time.

Transaction pool and block production will not really compete for I/O. The transaction pool will very likely only use stuff that is already in cache or only will read stuff from the db that will go into the cache afterwards. Block production will use the same cache. So, they probably compete about CPU resources, but probably not I/O.

**Comment by NikVolf** [comment]
Maybe, but not necessary, those transaction from network/rpc can hit completely different storage items compared to ones that constitute the block.

**Comment by bkchr** [comment]
> Maybe, but not necessary, those transaction from network/rpc can hit completely different storage items compared to ones that constitute the block.

That is right, but only a very limited set of different storage items as the validation is very lightweight. 

Do you have observed any problems in your benchmarks that the validation slows down the block production?

**Comment by NikVolf** [comment]
I am not yet sure that I can attribute any slowdowns to this, It's rather hard to check.

I will keep this issue updated.
